pkg/cloud/rgraph/exec/executor_parallel.go

```
@@ -0,0 +1,172 @@
+*/
+
+package exec
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud"
+	"github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud/rgraph/algo"
+	"k8s.io/klog/v2"
+)
+
+func defaultParallelExecutorConfig() *ExecutorConfig {
+	return &ExecutorConfig{
+		DryRun:        false,
+		ErrorStrategy: ContinueOnError,
+		Timeout:       5 * time.Minute,
+	}
+}
+
+// NewParallelExecutor returns a new Executor that runs tasks multi`-threaded.
+func NewParallelExecutor(pending []Action, opts ...Option) (*parallelExecutor, error) {
+	ret := &parallelExecutor{
+		config: defaultParallelExecutorConfig(),
+		result: &Result{Pending: pending},
+		pq:     algo.NewParallelQueue[Action](),
+	}
+	for _, opt := range opts {
+		opt(ret.config)
+	}
+
+	if err := ret.config.validate(); err != nil {
+		return nil, err
+	}
+	return ret, nil
+}
+
+type parallelExecutor struct {
+	config *ExecutorConfig
+	c      cloud.Cloud
+
+	// lock guards results
+	lock   sync.Mutex
+	result *Result
+
+	pq   *algo.ParallelQueue[Action]
+	done chan *TraceEntry
+}
+
+var _ Executor = (*parallelExecutor)(nil)
+
+func (ex *parallelExecutor) Run(ctx context.Context, c cloud.Cloud) (*Result, error) {
+	ex.executeTasks()
+	subctx, cancel := context.WithTimeout(ctx, ex.config.Timeout)
+
+	runAction := func(ctx context.Context, a Action) error {
+		te := &TraceEntry{
+			Action: a,
+			Start:  time.Now(),
+		}
+		klog.V(4).Infof("Run action %s", a)
+		events, runErr := a.Run(ctx, c)
+		te.End = time.Now()
+		klog.V(4).Infof("Finish action %s, err: %v", a, runErr)
+
+		ex.addActionResult(a, runErr)
+
+		if runErr != nil {
+			klog.Infof("Got error  %v, from action %s error_strategy: %s", runErr, a, ex.config.ErrorStrategy)
+			// check error strategy and decide if new actions should be executed.
+			if ex.config.ErrorStrategy == StopOnError {
+				return fmt.Errorf("parallelExecutor: stopping execution for Action %s (got %v)", a, runErr)
```

Do we need to cancel the context to see if we can make the existing operations quit early?

---

context is canceled in `(ex *parallelExecutor) Run` line 115

---

This swallows the error -- you should chain with %w.

(ignore my comment about cancel())

---

done